### PR TITLE
Bugfixes in SVHN and DenseDesignMatrixPyTables

### PR DESCRIPTION
**Pull request by sandeepkbhat**
* [**fix**] split_train_valid method in SVHN class is expected to return two tuples, each with two values. But it was returning 4 values. This was
      wrongly introduced during commit af07b6cc076942ef4f80c84ec7c7148afb3999ad ("remove one_hot, first try...")
* [**fix**] Added support to specify y_labels in DenseDesignMatrixPyTables constructor.This should have been done as part of the following commits af07b6cc076942ef4f80c84ec7c7148afb3999ad ("remove one_hot, first try...") d6050106f137147ba976e472748ad77273218b55 ("Modified all super() params...")
* [**fix**] Table shape for y (the labels) is now set to [sizes[which_set], 1].
      Before it was being wrongly set to [sizes[which_set], 10]. Also y
      labels for SVHN now range from [0,9], similar to the MNIST dataset
      labels. These changes should have been done as part of commit
      af07b6cc076942ef4f80c84ec7c7148afb3999ad ("remove one_hot, first try...")
* [**fix**] Now the SVHN dataset class properly sets the filters attribute
      before call to resize().
* [**chg**] Removed hardcoded title specification in init_hdf5(). It can now
      be specified by an optional argument. SVHN class now sets the optional
      title argument.
* [**chg**] If dataset y is of integer type then int type atom is used by
      init_hdf5() for storing the labels in pytables.
* [**new**] Added ./pylearn2/scripts/datasets/download_svhn.sh script for SVHN
      dataset download to the $PYLEARN2_DATA_PATH/SVHN/format2 directory. The
      script is closely based on download_cifar10.sh, but it uses a verbose
      wget or curl.
* [**new**] Added ./pylearn2/scripts/datasets/make_svhn_pytables.py which
      converts the download MAT files for SVHN dataset to Pytables format.
* [**note**] The above fixes have been tested using SVHN YAML files in ./pylearn2/scripts/papers/maxout as follows:
       * [x] Setup your theano/pylearn2 to use GPU and set config.floatX = 'float32'
       * [x] Set the `PYLEARN2_DATA_PATH` environment variable.
       * [x] Download the SVHN dataset MAT files by running
           `./download_svhn.sh`
       * [x] Generate the necessary Pytables files by running
           `python make_svhn_pytables.py`
       * [x] Copy the generated dataset to an alternate path say /tmp/SVHN
       * [x] Set the `SVHN_LOCAL_PATH` environment variable to this path
          `export SVHN_LOCAL_PATH=/tmp/SVHN/`
       * [x] Pre-process the dataset by running
           `python svhn_preprocessing.py`
       * [x] Train on the processed dataset by running
           `THEANO_FLAGS="device=gpu,floatX=float32" python train.py svhn.yaml`
       * [x] For the first 10 iterations I got the following test_y_misclass
           1. test_y_misclass: 0.923491537571
           2. test_y_misclass: 0.804302871227
           3. test_y_misclass: 0.0927878990769
           4. test_y_misclass: 0.0584590695798
           5. test_y_misclass: 0.0481450334191
           6. test_y_misclass: 0.0444889366627
           7. test_y_misclass: 0.0405634306371
           8. test_y_misclass: 0.0373691543937
           9. test_y_misclass: 0.0350215472281
           10. test_y_misclass: 0.0334051698446

### CODE REVIEW

**Comment by memimo** [comment]
Thanks for the PR, I think you are right about one_hot, someone broke it. I don't have time to review it in the next week. But in meantime, look at #892 too as it is addressing some other issues too, and ideally one PR should not break the other.

**Comment by sandeepkbhat** [comment]
@memimo : I am not sure if my PR is 100% compatible with #892 . If you see any issues during your review let me know and I can work on fixing those. Also, looks like the Travis CI check failed on my PR, mostly due to Pep8 and pyflakes issues in a newly added file. I am not exactly sure on how to proceed on this issue. Let me know if you have any suggestions. Thanks.

**Comment by memimo** [comment]
You can use autopep8 and pyflakes to solve their issues. These two links are good sources for PR:
http://deeplearning.net/software/pylearn2/internal/dev_start_guide.html#dev-start-guide
http://deeplearning.net/software/pylearn2/internal/pull_request_checklist.html#pull-request-checklist

As for PR #892. Probably, we should merge that first, and then you rebase you PR on top of that.

File: pylearn2/datasets/dense_design_matrix.py [hunk 1]
```diff
@@ -1205,17 +1214,27 @@ def init_hdf5(self, path, shapes):
...
         # make pytables
         ensure_tables()
-        h5file = tables.openFile(path, mode="w", title="SVHN Dataset")
+        h5file = tables.openFile(path, mode="w", title=title)
         gcolumns = h5file.createGroup(h5file.root, "Data", "Data")
         atom = (tables.Float32Atom() if config.floatX == 'float32'
                 else tables.Float64Atom())
         h5file.createCArray(gcolumns, 'X', atom=atom, shape=x_shape,
                             title="Data values", filters=self.filters)
+        if y_dtype != 'float':
```
Comment (lamblin) [inline pylearn2/datasets/dense_design_matrix.py]: Can you make sure that `y_dtype` is either `'float'` or `'int'` here? It may be confusing if a user passes `y_dtype = 'int32'` for instance,

File: pylearn2/scripts/datasets/make_svhn_pytables.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+from theano import config
+from pylearn2.datasets import preprocessing
+from pylearn2.datasets.svhn import SVHN
+from pylearn2.utils.string_utils import preprocess
+
+orig_path  = preprocess('${PYLEARN2_DATA_PATH}/SVHN/format2')
+
+# Check if MAT files have been downloaded
+if not os.path.isdir(orig_path):
+    raise IOError("You need to download the SVNH format2 dataset MAT files "
```
Comment (lamblin) [inline pylearn2/scripts/datasets/make_svhn_pytables.py]: It should be "SVHN" instead of "SVNH" in a couple of places.

File: pylearn2/scripts/datasets/make_svhn_pytables.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+print """
+      ***************************************************************
+      Please ignore the warning produced during this MAT -> Pytables
+      conversion for the SVNH dataset. If you are creating the
+      pytables for the first time then no files are modified/over-written,
+      they are simply written for the first time.
+      ***************************************************************
+      """
+
+test = SVHN('test', path=local_path+'/')
```
Comment (lamblin) [inline pylearn2/scripts/datasets/make_svhn_pytables.py]: Why the added `/` here? If we need to combine paths, we should avoid hardcoding OS-specific strings.

**Comment by dwf** [comment]
Build failure...

**Comment by sandeepkbhat** [comment]
Following is the error in both 2.6 and 3.4 Python on Travis CI
> AssertionError: Docstring format not respected:
> scripts/datasets/make_svhn_pytables.py failed to run so format cannot be checked. Error message:
> You need to define your PYLEARN2_DATA_PATH environment variable. If you are
> using a computer at LISA, this should be set to /data/lisa/data.

I think the reason it does not run is because the mat files needed for creating SVHN dataset do not exist on Travis CI. I am not sure what I can do to fix this. 

I have another PR (#1340) which downloads the SVHN dataset mat files (the file download_svhn.sh
is present in this PR also). But it needs to be run on Travis CI before the make_svhn_pytables.py script can work.

**Comment by lamblin** [comment]
I don't think we want to download SVHN on Travis each time.
You can skip the test (raise SkipTest) if PYLEARN2_DATA_PATH is not defined, there are examples in other files.

**Comment by sandeepkbhat** [comment]
I still see that the label on this is "Submitter Action Required". I am not sure what that action is. Please let me know if any additional work is needed from my side to enable merging of this PR to the master branch. Thanks.

**Comment by lamblin** [comment]
Sorry, GitHub does not notify us when a PR is updated, only when you comment on it, so I had no idea that you submitted new commits.
Also, if you could rebase and get rid of the 3 merge commits, that would be appreciated.
@memimo can you please review it?

**Comment by sandeepkbhat** [comment]
Thanks @lamblin for that reponse. I did not quite get the rebase comment. I have already merged the latest pylearn2 code onto my pytables_svhn_bugfixes branch in the above PR. What should I rebase it to, the latest pylearn2 code? Could you please expand on that. Thank you.

**Comment by lamblin** [comment]
There are two ways to deal with conflicts: merge, or rebase. Merging makes for a messier history, so if possible, we prefer the PRs being rebased on top of the latest "master" branch of lisa-lab/pylearn2.
Then, since rebasing rewrites the history, you would need to use the `-force` option for `git push` when updating the PR.

**Comment by sandeepkbhat** [comment]
@lamblin I have rebased my PR on top of the pylearn2 master. Please review and merge it into master. Thanks.

File: pylearn2/datasets/dense_design_matrix.py (hunk 1 above)
Comment (lamblin) [inline pylearn2/datasets/dense_design_matrix.py]: (Done.)

File: pylearn2/scripts/datasets/make_svhn_pytables.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+Script to split the downloaded SVHN .mat files to 'test', 'valid', and 'train'
+sets and save them as pytables. Ensure that you have set the PYLEARN2_DATA_PATH
+environment variable and downloaded the .mat files to the path
+${PYLEARN2_DATA_PATH}/SVHN/format2/
+"""
+import os
+from pylearn2.datasets.svhn import SVHN
+from pylearn2.utils.string_utils import preprocess
+
+assert 'PYLEARN2_DATA_PATH' in os.environ, "PYLEARN2_DATA_PATH not defined"
```
Comment (lamblin) [inline pylearn2/scripts/datasets/make_svhn_pytables.py]: Could you wrap everything but "import" statements into `if __name__ == '__main__'`?

File: pylearn2/devtools/tests/test_format.py [hunk 5]
```diff
@@ -556,6 +556,7 @@
     'scripts/datasets/make_cifar10_gcn_whitened.py',
     'scripts/datasets/make_cifar100_patches.py',
     'scripts/datasets/make_cifar100_gcn_whitened.py',
+    'scripts/datasets/make_svhn_pytables.py',
```
Comment (lamblin) [inline pylearn2/devtools/tests/test_format.py]: Why did you need to add it to the whitelist? We are not supposed to add new files in there.

**Comment by lamblin** [comment]
I'll merge because of the numerous bug fixes. It would be nice it you could take care of the last minor points in a subsequent PR.
Thanks!
